File: js/random-tree.js
'use strict';

const assert = require('assert');
const {BLAKE2b} = require('bcrypto');
const {Tree} = require('urkel');
const {randomStuffByte} = require('../lib/rand');
const fs = require('bfile');

const MAX_ITER = 40000;

const OP_COMMIT = 1;
const OP_INSERT = 2;
const OP_REVERT = 3;
const OP_UPDATE = 4;
const OP_REMOVE = 5;

function randBytes(nextRandByte, n) {
  const buf = Buffer.alloc(n, 0x00);

  for (let i = 0; i < n; i++)
    buf[i] = nextRandByte();

  return buf;
}

function randKey(nextRandByte) {
  const keySize = nextRandByte();
  const key = randBytes(nextRandByte, keySize);
  const hash = BLAKE2b.digest(key);

  return hash;
}

function randValue(nextRandByte) {
  const size = nextRandByte();
  return randBytes(nextRandByte, size);
}

function getDecision(nextRandByte) {
  const byte = nextRandByte();
  let prob = 0;

  // 10 %
  prob += 25;

  // around 10%
  if (byte <= prob)
    return OP_COMMIT;

  // around 10%
  prob += 25;

  if (byte <= prob)
    return OP_UPDATE;

  // around 5%
  prob += 13;

  if (byte <= prob)
    return OP_REVERT;

  // around 5%
  prob += 13;

  if (byte <= prob)
    return OP_REMOVE;

  // around 70 %
  return OP_INSERT;
}

/*
 * Important thing to manage between this and C is to
 * maintain order of rand operations.
 */

async function run(nextRandByte) {
  const tree = new Tree({
    hash: BLAKE2b,
    bits: 256,
    prefix: './tree'
  });
  const fd = await fs.open('./proofs', 'w');

  await tree.open();

  const roots = [];
  const keys = [];

  let txn = tree.txn();
  for (let i = 0; i < MAX_ITER; i++) {
    const decision = getDecision(nextRandByte);

    switch (decision) {
      case OP_INSERT: {
        const key = randKey(nextRandByte);
        const value = randValue(nextRandByte);
        await txn.insert(key, value);
        keys.push(key);
        break;
      }

      case OP_UPDATE: {
        const key = keys[nextRandByte() % keys.length];

        if (keys.length === 0)
          break;

        const value = randValue(nextRandByte);

        await txn.insert(key, value);
        break;
      }

      case OP_REMOVE: {
        const key = keys[nextRandByte() % keys.length];

        if (keys.length === 0)
          break;

        await txn.remove(key);
        break;
      }

      case OP_COMMIT: {
        const root = await txn.commit();
        roots.push(root);
        txn = tree.txn();
        break;
      }

      case OP_REVERT: {
        if (roots.length === 0)
          break;

        const rnum = nextRandByte();
        const root = roots[rnum % roots.length];
        await tree.inject(root);
        break;
      }

      default: {
        throw new Error('Unknown decision.');
      }
    }

    if (roots.length > 0 && keys.length > 0) {
      let rnum = nextRandByte();
      const key = keys[rnum % keys.length];
      rnum = nextRandByte();
      const root = roots[rnum % roots.length];
      const snap = tree.snapshot(root);
      const proof = await snap.prove(key);
      const proofRaw = proof.encode(BLAKE2b, 256);

      await fs.write(fd, proofRaw, 0, proofRaw.length);
    }
  }

  await txn.commit();
  await tree.close();
  await fs.close(fd);
}

(async () => {
  const seedArg = process.argv[2] || 2022;
  const seed = Number(seedArg);

  assert(!Number.isNaN(seed));
  const randGen = randomStuffByte(seed);
  const nextRandByte = () => {
    const {value, done} = randGen.next();
    assert(!done);
    return value;
  };

  await run(nextRandByte);
})().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: src/random-tree.c

#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <stdint.h>
#include <urkel.h>
#include <stdio.h>

#include "util.h"

#define MAX_ITER 40000
#define HASH_SIZE 32

#define OP_COMMIT 1
#define OP_INSERT 2
#define OP_REVERT 3
#define OP_UPDATE 4
#define OP_REMOVE 5

struct root_list {
  uint8_t roots[MAX_ITER][HASH_SIZE];
  uint32_t size;
};

struct key_list {
  uint8_t keys[MAX_ITER][HASH_SIZE];
  uint32_t size;
};

void
rand_bytes(int32_t *seed, uint8_t *data, uint8_t n) {
  int i;
  for (i = 0; i < n; i++)
    data[i] = random_stuff_byte(seed);
}

void
rand_key(int32_t *seed, uint8_t *hash) {
  uint8_t key_size = random_stuff_byte(seed);
  uint8_t key[256];
  rand_bytes(seed, key, key_size);
  urkel_hash(hash, key, key_size);
}

void
rand_value(int32_t *seed, uint8_t *value, uint8_t *size) {
  uint8_t vsize = random_stuff_byte(seed);

  rand_bytes(seed, value, vsize);
  *size = vsize;
}

uint8_t
get_decision(int32_t *seed) {
  uint8_t byte = random_stuff_byte(seed);
  uint8_t prob = 0;

  /* around 10% */
  prob += 25;

  if (byte <= prob)
    return OP_COMMIT;

  /* around 10% */
  prob += 25;

  if (byte <= prob)
    return OP_UPDATE;

  /* around 5% */
  prob += 13;

  if (byte <= prob)
    return OP_REVERT;

  /* around 5% */
  prob += 13;

  if (byte <= prob)
    return OP_REMOVE;

  /* around 70 % */
  return OP_INSERT;
}

void
run(int32_t *seed) {
  urkel_t *db;
  urkel_tx_t *tx;
  FILE *proof_fd;

  int i;
  struct root_list *roots = malloc(sizeof(struct root_list));
  struct key_list *keys = malloc(sizeof(struct key_list));
  memset(roots, 0, sizeof(struct root_list));
  memset(keys, 0, sizeof(struct key_list));

  proof_fd = fopen("./proofs", "w");
  assert(proof_fd != NULL);

  db = urkel_open("./tree");
  assert(db != NULL);

  tx = urkel_tx_create(db, NULL);
  assert(tx != NULL);

  for (i = 0; i < MAX_ITER; i++) {
    uint8_t decision = get_decision(seed);

    switch (decision) {
      case OP_INSERT: {
        uint8_t key_hash[HASH_SIZE] = {0};
        uint8_t value[256];
        uint8_t val_size = 0;

        rand_key(seed, key_hash);
        rand_value(seed, value, &val_size);
        assert(urkel_tx_insert(tx, key_hash, value, val_size));
        memcpy(keys->keys[keys->size], key_hash, HASH_SIZE);
        keys->size++;

        break;
      }

      case OP_UPDATE: {
        uint8_t rnum = random_stuff_byte(seed);
        uint8_t key_hash[HASH_SIZE] = {0};
        uint8_t value[256];
        uint8_t val_size = 0;

        if (keys->size == 0)
          break;

        memcpy(key_hash, keys->keys[rnum % keys->size], HASH_SIZE);
        rand_value(seed, value, &val_size);
        assert(urkel_tx_insert(tx, key_hash, value, val_size));
        break;
      }

      case OP_REMOVE: {
        uint8_t rnum = random_stuff_byte(seed);
        uint8_t key_hash[HASH_SIZE] = {0};

        if (keys->size == 0)
          break;

        memcpy(key_hash, keys->keys[rnum % keys->size], HASH_SIZE);

        if (!urkel_tx_has(tx, key_hash))
          break;

        assert(urkel_tx_remove(tx, key_hash));
        break;
      }

      case OP_COMMIT: {
        assert(urkel_tx_commit(tx));
        urkel_tx_destroy(tx);
        tx = urkel_tx_create(db, NULL);
        urkel_root(db, roots->roots[roots->size]);
        roots->size++;
        assert(tx != NULL);
        break;
      }
      case OP_REVERT: {
        uint8_t rnum;
        uint8_t *root;

        if (roots->size == 0)
          break;

        rnum = random_stuff_byte(seed);
        root = roots->roots[rnum % roots->size];
        urkel_inject(db, root);
        break;
      }
      default: {
        assert(0);
      }
    }

    {
      uint8_t *proof, *key, *root;
      uint8_t rnum;
      size_t proof_len;

      if (roots->size > 0 && keys->size > 0) {
        rnum = random_stuff_byte(seed);
        key = keys->keys[rnum % keys->size];
        rnum = random_stuff_byte(seed);
        root = roots->roots[rnum % roots->size];

        urkel_prove(db, &proof, &proof_len, key, root);
        fwrite(proof, 1, proof_len, proof_fd);
        free(proof);
      }
    }
  }

  assert(urkel_tx_commit(tx));

  free(roots);
  free(keys);
  urkel_tx_destroy(tx);
  urkel_close(db);
  fclose(proof_fd);
}

int main(int argc, char **argv) {
  int32_t seed = 2022;

  if (argc > 1)
    seed = atoi(argv[1]);


  run(&seed);

  return 0;
}
